internal/services/notification/db/db.go

```diff
@@ -0,0 +1,122 @@
+package db
+
+import (
+	"context"
+	stdsql "database/sql"
+
+	sq "github.com/huandu/go-sqlbuilder"
+	"github.com/rs/zerolog"
+	"github.com/sorintlab/errors"
+
+	"agola.io/agola/internal/services/notification/db/objects"
+	"agola.io/agola/internal/sqlg"
+	"agola.io/agola/internal/sqlg/sql"
+	"agola.io/agola/services/notification/types"
+)
+
+//go:generate ../../../../tools/bin/dbgenerator -type db -component notification
+
+type DB struct {
+	log zerolog.Logger
+	sdb *sql.DB
+}
+
+func NewDB(log zerolog.Logger, sdb *sql.DB) (*DB, error) {
+	return &DB{
+		log: log,
+		sdb: sdb,
+	}, nil
+}
+
+func (d *DB) DBType() sql.Type {
+	return d.sdb.Type()
+}
+
+func (d *DB) Do(ctx context.Context, f func(tx *sql.Tx) error) error {
+	return errors.WithStack(d.sdb.Do(ctx, f))
+}
+
+func (d *DB) ObjectsInfo() []sqlg.ObjectInfo {
+	return objects.ObjectsInfo
+}
+
+func (d *DB) Flavor() sq.Flavor {
+	switch d.sdb.Type() {
+	case sql.Postgres:
+		return sq.PostgreSQL
+	case sql.Sqlite3:
+		return sq.SQLite
+	}
+
+	return sq.PostgreSQL
+}
+
+func (d *DB) exec(tx *sql.Tx, rq sq.Builder) (stdsql.Result, error) {
+	q, args := rq.BuildWithFlavor(d.Flavor())
+	// d.log.Debug().Msgf("q: %s, args: %s", q, util.Dump(args))
+
+	r, err := tx.Exec(q, args...)
+	return r, errors.WithStack(err)
+}
+
+func (d *DB) query(tx *sql.Tx, rq sq.Builder) (*stdsql.Rows, error) {
+	q, args := rq.BuildWithFlavor(d.Flavor())
+	// d.log.Debug().Msgf("start q: %s, args: %s", q, util.Dump(args))
+
+	r, err := tx.Query(q, args...)
+	// d.log.Debug().Msgf("end q: %s, args: %s", q, util.Dump(args))
+	return r, errors.WithStack(err)
+}
+
+func mustSingleRow[T any](s []*T) (*T, error) {
+	if len(s) > 1 {
+		return nil, errors.Errorf("too many rows returned")
+	}
+	if len(s) == 0 {
+		return nil, nil
+	}
+
+	return s[0], nil
+}
+
+func (d *DB) GetRunWebhookDeliverysByDeliveryStatus(tx *sql.Tx, deliveryStatus []types.DeliveryStatus, limit int) ([]*types.RunWebhookDelivery, error) {
+	q := runWebhookDeliverySelect().OrderBy("sequence").Asc()
+	if len(deliveryStatus) > 0 {
+		q.Where(q.In("delivery_status", sq.Flatten(deliveryStatus)...))
+	}
+
+	if limit > 0 {
+		q.Limit(limit)
+	}
+
+	runWebhookDeliverys, _, err := d.fetchRunWebhookDeliverys(tx, q)
+	return runWebhookDeliverys, errors.WithStack(err)
+}
+
+func (d *DB) GetRunWebhookDeliverys(tx *sql.Tx, limit int) ([]*types.RunWebhookDelivery, error) {
+	return d.GetRunWebhookDeliverysByDeliveryStatus(tx, []types.DeliveryStatus{}, limit)
+}
+
+func (d *DB) GetRunWebhookDeliveryByID(tx *sql.Tx, runWebhookDeliveryID string) (*types.RunWebhookDelivery, error) {
+	q := runWebhookDeliverySelect()
+	q.Where(q.E("id", runWebhookDeliveryID))
+	runWebhookDeliverys, _, err := d.fetchRunWebhookDeliverys(tx, q)
```

Comment:
the generated functions could have the wrong plural forms (since it only adds an s) but the manually typed code should use the right plural form (deliveries)